Cmdline/Action/Prompt.cs

```diff
@@ -55,6 +63,121 @@ public int RunCommand(GameInstanceManager manager, object raw_options)
+            catch (InvalidOperationException)
+            {
+                // InvalidOperationException is thrown in a mintty on Windows
+                return Console.ReadLine();
+            }
+        }
+
+        private string[] GetSuggestions(string text, int index)
+        {
+            string[]     pieces = text.Split(new char[] { ' ' });
+            TypeInfo     ti     = typeof(Actions).GetTypeInfo();
+            List<string> extras = new List<string> { exitCommand, "help" };
+            foreach (string piece in pieces.Take(pieces.Length - 1))
+            {
+                PropertyInfo pi = ti.DeclaredProperties
+                    .FirstOrDefault(p => p.GetCustomAttribute<VerbOptionAttribute>().LongName == piece);
+                if (pi == null)
+                {
+                    // Couldn't find it, no suggestions
+                    return null;
+                }
+                ti = pi.PropertyType.GetTypeInfo();
+                extras.Clear();
+            }
+            var lastPiece = pieces.LastOrDefault() ?? "";
+            return lastPiece.StartsWith("--") ? GetOptions(ti, lastPiece.Substring(2))
+                : HasVerbs(ti)                ? GetVerbs(ti, lastPiece, extras)
+                : WantsAvailIdentifiers(ti)   ? GetAvailIdentifiers(lastPiece)
+                : WantsInstIdentifiers(ti)    ? GetInstIdentifiers(lastPiece)
+                : WantsGameInstances(ti)      ? GetGameInstances(lastPiece)
+                :                               null;
+        }
+
+        private string[] GetOptions(TypeInfo ti, string prefix)
+        {
+            return ti.DeclaredProperties
+                .Select(p => p.GetCustomAttribute<OptionAttribute>()?.LongName)
+                .Where(o => o != null && o.StartsWith(prefix, StringComparison.InvariantCultureIgnoreCase))
+                .OrderBy(o => o)
+                .Select(o => $"--{o}")
+                .ToArray();
+        }
+
+        private bool HasVerbs(TypeInfo ti)
+        {
+            return ti.DeclaredProperties
+                .Any(p => p.GetCustomAttribute<VerbOptionAttribute>() != null);
+        }
+
+        private string[] GetVerbs(TypeInfo ti, string prefix, IEnumerable<string> extras)
+        {
+            return ti.DeclaredProperties
+                .Select(p => p.GetCustomAttribute<VerbOptionAttribute>()?.LongName)
+                .Where(v => v != null)
+                .Concat(extras)
+                .Where(v => v.StartsWith(prefix, StringComparison.InvariantCultureIgnoreCase))
+                .OrderBy(v => v)
+                .ToArray();
+        }
+
+        private bool WantsAvailIdentifiers(TypeInfo ti)
+        {
+            return ti.DeclaredProperties
+                .Any(p => p.GetCustomAttribute<AvailableIdentifiersAttribute>() != null);
+        }
+
+        private string[] GetAvailIdentifiers(string prefix)
+        {
+            CKAN.GameInstance inst = MainClass.GetGameInstance(manager);
```

Comment:
This throws a `NoGameInstanceKraken` if there is no default instance and also non selected with `--instance` / `--gamedir`, and makes the prompt close and exit with code 1.
```
CKAN v1.30.5> install AsI don't know where a game instance is installed.
Use 'ckan instance help' for assistance in setting this.

Process finished with exit code 1.
```

We should catch it and just write the message

Reply:
Good catch, `upgrade` has the same problem through a different code path. I'm going to look for a solution that handles both `GetAvailIdentifiers` and `GetInstIdentifiers` consistently...

Reply:
A couple of findings...

- If we just `Console.WriteLine` in the completion handler and return an empty array, it messes up the display (at least on Windows); the `ReadLine` code doesn't know that the cursor has moved, so the user is still editing the string but can't see it
- There doesn't seem to be any way to tell `ReadLine` that this has happened
- Given the above, the best option seems to be to catch the exception after it interrupts the `ReadLine` call, so we can start a new `ReadLine` call after that with fresh state
- If we do that, we can't restore the previously entered comment, even if we capture it via the exceptions; there's no way to pass it back to `ReadLine` as the initial command to edit. However, maybe that's desirable, as there is no way to finish an `install` command successfully if you have no game instance

Reply:
Pushed a commit that should improve this:

![image](https://user-images.githubusercontent.com/1559108/164769577-19842f69-d2f0-45ce-ada6-acc2ee6cd92a.png)